### DISCUSSION TITLE
Dependency Injection Into Business Rule

### DISCUSSION THREAD

**Discussion by saulyt** [MarimerLLC/csla#2063] [Questions] [answered]
Is there a way to inject a dependency into a custom business rule? I am looking for any way, either via constructor, parameter, and if those are not possible than getting the dependency via the IServiceProvider. (I can't figure out how to access the IServiceProvider within the rule.) 

Thank you

**Comment by rockfordlhotka** [comment, marked as answer]
That is not possible in CSLA 5, but should be possible in CSLA 6 (though that is something I'd need to confirm).

Out of curiosity, what do you want to inject into the rule instance?

**Reply by saulyt** [reply]
I have an object that I am using to retrieve lookup lists of key/values. ( The lists themselves are actually ReadOnlyLists that implement an interface for key and value.) The values for various lists are loaded from the DB ( via DataPortal) into MemoryCache the first time they are accessed via this object. Subsequent access is retrieved from the cache. I would like to access this object inside the business rule so that I can check that a user entered value is in the list. The object itself has a constructor parameter for the memory cache which is configured in DI container at app startup

Thanks

**Reply by saulyt** [reply]
I wound up putting my caching logic into the list objects fetch methods  and injecting the IMemoryCache into the Fetch methods..

**Reply by rockfordlhotka** [reply]
I usually create a `static` cache for these things and use them from a global "App" type or something.

I really like the idea of injecting a singleton into a rule though, as that would be much cleaner, or at least more modern.
